kaiax/supply/impl/init.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2024 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package supply
+
+import (
+	"math/big"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/kaiachain/kaia/accounts/abi/bind/backends"
+	"github.com/kaiachain/kaia/kaiax/reward"
+	"github.com/kaiachain/kaia/kaiax/supply"
+	"github.com/kaiachain/kaia/log"
+	"github.com/kaiachain/kaia/params"
+	"github.com/kaiachain/kaia/storage/database"
+)
+
+var (
+	_ supply.SupplyModule = &SupplyModule{}
+
+	logger = log.NewModuleLogger(log.KaiaxSupply)
+
+	// A day; Some total supply consumers might want daily supply.
+	// It may be evicted by the historic query though.
+	supplyCacheSize       = 86400
+	checkpointInterval    = uint64(128)    // AccReward checkpoint interval
+	accumulateLogInterval = uint64(102400) // Periodic log in accumulateRewards().
+)
+
+func nearestCheckpointNumber(num uint64) uint64 {
+	return num - (num % checkpointInterval)
+}
+
+type InitOpts struct {
+	ChainKv      database.Database
+	ChainConfig  *params.ChainConfig
+	Chain        backends.BlockChainForCaller
+	RewardModule reward.RewardModule
+}
+
+type SupplyModule struct {
+	InitOpts
+
+	// Accumulated supply checkpoint so far.
+	// This in-memory variables advance every block, but the database is updated every checkpointInterval.
+	mu            sync.RWMutex
+	lastAccNum    uint64            // Last AccReward number
+	lastAccReward *supply.AccReward // Last AccReward
+
+	// Stops long-running tasks.
+	quit   uint32         // stops the synchronous loop in accumulateCheckpoint
+	quitCh chan struct{}  // stops the goroutine in select loop
+	wg     sync.WaitGroup // wait for the goroutine to finish
+
+	supplyCache *lru.ARCCache // (number uint64) -> (totalSupply *TotalSupply)
```

Comment:
What's the purpose of `supplyCache`? It only stores `totalSupply` when there's API request, which means full node can't get the `totalSupply` from cache if hadn't. So would we expect the node operator to call `GetTotalSupply` for every block to cache the data? If not, cache won't be filled for block after `TrieInMemory` in case of full node.

Also, I think introducing `accRewardCache` can be effective.

Reply:
It is for caching the API result. I found out that caching accReward won't help full nodes anyway because CanonicalBurn data is unavailable after TriesInMemory. So I removed accRewardCache.

Reply:
Having accRewardCache might speed up GetTotalSupply(128k + r) cases, but even without cache it only iterates up to 128 headers which won't take too long.

Reply:
Ok. I also think 127 iterations won't be matter.

I just wanted to confirm that it was intended and it seems so. And for full node, it can still serve blocks with `num%128==0`, which only makes acceptable accuracy errors.

Reply:
You're right. Full nodes can serve blocks with `num%128=0` in which case there exists ReadAccReward(). Because DB has its own cache, we don't need separate cache.